lib/spack/spack/util/windows_registry.py

```diff
@@ -0,0 +1,262 @@
+    the root key used to instantiate this class.
+    """
+
+    def __init__(self, key, root_key=HKEY_CONSTANT("HKEY_CURRENT_USER")):
+        """Constructs a Windows Registry entrypoint to key provided
+        root_string should be an already open root key or an hkey constant if provided
+
+        Args:
+            key (str): registry key to provide root for registry key for this clas
+            root_string: Already open registry key or HKEY constant to provide access into
+                         the Windows registry. Registry access requires an already open key
+                         to get an entrypoint, the HKEY constants are always open, or an already
+                         open key can be used instead.
+        """
+        if not is_windows:
+            raise RuntimeError(
+                "Cannot instantiate Windows Registry class on non Windows platforms"
+            )
+        self.key = key
+        self.root = root_key
+        try:
+            self.reg = RegistryKey(
+                os.path.join(str(self.root), self.key),
+                winreg.OpenKeyEx(self.root.hkey, self.key, access=winreg.KEY_READ),
+            )
+        except FileNotFoundError as e:
+            if e.winerror == 2:
+                self.reg = None
+            else:
+                raise e
+
+    def get_value(self, value_name):
+        return self.reg.get_value(value_name)
+
+    def get_subkey(self, subkey_name):
+        return self.reg.get_subkey(subkey_name)
+
+    def _traverse_subkeys(self, stop_condition):
+        """Perform simple BFS of subkeys, returning the key
+        that successfully triggers the stop condition.
+        Args:
+            stop_condition: lambda or function pointer that takes a single argument
+                            a key and returns a boolean value based on that key
+        Return:
+            the key if stop_condition is triggered, or None if not
+        """
+        queue = self.reg.subkeys
+        for key in queue:
+            if stop_condition(key):
+                return key
+            queue.extend(key.subkeys)
+        return None
+
+    def find_subkey(self, subkey_name, recursive=True):
+        """If non recursive, this method is the same as get subkey with error handling
+        Otherwise perform a BFS of subkeys until desired key is found
+        Returns None or RegistryKey object corresponding to requested key name
+
+        Args:
+            subkey_name (str): string representing subkey to be searched for
+            recursive (bool): optional argument, if True, subkey need not be a direct
+                                sub key of this registry entry, and this method will
+                                search all subkeys recursively.
+                                Default is True
+        Return:
+            the desired subkey as a RegistryKey object, or none
+        """
+        if not recursive:
+            try:
+                self.get_subkey(subkey_name)
```

Comment:
.... I'll fix that.

I had intended this top level class to be more of the actual usable interface and the `RegistryKey` to be more of a raw interface for interacting with the winreg module directly, so I left `get_subkey` and `get_value` to fail in case an interface wanted to handle that case differently at some point.